src/LuaPiGui.cpp

```diff
@@ -1275,31 +1365,215 @@ TScreenSpace lua_world_space_to_screen_space(const Body *body)
+	else if (a_moon && b_moon)
+		result = body->GetLabel() < other->GetLabel();
+	// a planet is larger than any non-planet
+	else if (a_planet)
+		result = true;
+	// a non-planet is smaller than any planet
+	else if (b_planet)
+		result = false;
+	// a moon is larger than any non-moon
+	else if (a_moon)
+		result = true;
+	// a non-moon is smaller than any moon
+	else if (b_moon)
+		result = false;
+	// spacestation > city > ship > hyperspace cloud > cargo body > missile > projectile
+	else if (a == Object::Type::SPACESTATION)
+		result = true;
+	else if (b == Object::Type::SPACESTATION)
+		result = false;
+	else if (a == Object::Type::CITYONPLANET)
+		result = true;
+	else if (b == Object::Type::CITYONPLANET)
+		result = false;
+	else if (a == Object::Type::SHIP)
+		result = true;
+	else if (b == Object::Type::SHIP)
+		result = false;
+	else if (a == Object::Type::HYPERSPACECLOUD)
+		result = true;
+	else if (b == Object::Type::HYPERSPACECLOUD)
+		result = false;
+	else if (a == Object::Type::CARGOBODY)
+		result = true;
+	else if (b == Object::Type::CARGOBODY)
+		result = false;
+	else if (a == Object::Type::MISSILE)
+		result = true;
+	else if (b == Object::Type::MISSILE)
+		result = false;
+	else if (a == Object::Type::PROJECTILE)
+		result = true;
+	else if (b == Object::Type::PROJECTILE)
+		result = false;
+	else
+		Error("don't know how to compare %i and %i\n", a, b);
+
+	return result;
+}
+
+static int l_pigui_get_projected_bodies_grouped(lua_State *l)
 {
 	PROFILE_SCOPED()
-	LuaTable result(l);
+	//printf("l_pigui_get_projected_bodies_grouped\n");
+	const vector2d gap = LuaPull<vector2d>(l, 1);
+	std::vector<TScreenSpace> filtered;
+	filtered.reserve(Pi::game->GetSpace()->GetNumBodies());
+
 	for (Body *body : Pi::game->GetSpace()->GetBodies()) {
 		if (body == Pi::game->GetPlayer()) continue;
 		if (body->GetType() == Object::PROJECTILE) continue;
+		const TScreenSpace res = lua_world_space_to_screen_space(body); // defined in LuaPiGui.cpp
+		if (!res._onScreen) continue;
+		filtered.emplace_back(res);
+		filtered.back()._body = body;
+	}
+
+	std::vector<std::vector<TScreenSpace>> groups;
+	groups.reserve(filtered.size());
+
+	// Perform half-matrix double for
+	for (std::vector<TScreenSpace>::iterator it = filtered.begin(); it != filtered.end(); ++it) {
+		std::vector<TScreenSpace> group;
+		group.reserve(filtered.end() - it + 1);
+		// First element should always be the "media";
+		// so push twice this element:
+		//printf("Body 1 = %s (%f, %f)\n", (*it)._body->GetLabel().c_str(), (*it)._screenPosition.x, (*it)._screenPosition.y);
+		group.push_back((*it));
+		// Zeroed body so you could distinguish between a "media" element and a "real" element
+		group.back()._body = nullptr;
+		group.push_back((*it));
+
+		// Find near displayed bodies in remaining list of bodies
+		std::vector<TScreenSpace>::reverse_iterator current = std::reverse_iterator<std::vector<TScreenSpace>::iterator>(it + 1);
```

Comment:
I get a crash here, which I'm just investigating. 
When playing (Mars start) I take off and start flying straight ahead, then I look around with the middle mouse button

Reply:
It seems your S.O. is more susceptible to memory leaks, so I hope
you will do another test-fly...

Reply:
I've done a lot of test flights, with latest code, and I still get the problem.
It's a 100% repro' rate when using VS2017 x64|PreRelease (_which has Debugging & light optimisations_)

Reply:
It runs in Release|x64 but this means no-one will be able to run debug/prerelease builds using VS201X. It's definitely not correct.

Reply:
Well, I will investigate this issue, np.

I wonder if there's a way to get the same behaviour with with gcc...

Reply:
Great, thankyou

Reply:
@fluffyfreak , I'm wondering if you have time and will to collect some feedback on this issue:
may I ask you if you could uncomment the "printf" lines on this function and send me the output?

Reply:
I think the problem is that pop_back() technically invalidates any iterators pointing to the last element (or the end), and it2 can potentially be pointing to the final element when it's popped. Real vector implementations probably wouldn't invalidate the iterator, but MSVC debug will complain.

I couldn't get this working without switching to forward iterators in addition to the obvious solution of moving the increment inside the loop. I'm guessing that reverse iterators refer to the previous element internally, and so get invalidated if they're one-off-the-back too.

Reply:
I was thinking more or less the same, but blaming the `std::swap` which is moving/copying a previously moved (thus invalidated) iterator.

Anyway I pushed a "debug" version with (useful?) messages and a not "swap-and-pop" workaround.

Another solution, if the problem is on `pop_back()`, is [erase](http://www.cplusplus.com/reference/vector/vector/erase/) which return a valid iterator.

Reply:
I actually removed the std::move on sight, and forgot about it. It still needed both of the other modifications (forward iterator, decrement it2 before pop), so I don't know if the std::move is technically a problem or not.

Erase wouldn't necessarily return the correct iterator for the algorithm. it2 may or may not point to the last element.

Your copy-instead-of-swap code has the same iterator invalidation crash, with or without the std::move.

Reply:
Sorry for the late and thank for your support, I'm a little under pressure these days...

I would try a version in which `it2` increment is inside the loop (as per your suggestion) and the order of things is:
1) copy on `it2` (`(*it2) = filtered.back();`)
2) increment `it2` (which is "point `it2 `to previous element") (`it2++`)
3) `pop_back()`

This way, If I understand the problem correctly, `it2` will never point past-the-end and thus will never be invalidated... And the problem should be solved...

But I would add that the previous version (the one with forward iterator) was so unnatural and so full of checks and workaround that I prefer to propose a different route:
if the problem is as outlined above, to me a possible solution/workaround is to disable
those smart-iterator around the blamed code as explained [here](https://docs.microsoft.com/en-us/cpp/standard-library/debug-iterator-support?view=vs-2019).

Note we are working around not for a "bug" but for something less...

Reply:
Uploaded but not tested (I used some time at work), it should work anyway.

NOTE:

> forward iterator, decrement it2 before pop

equals to "increment reverse it before pop": So I did what you suggest, though not exactly

Reply:
Crashes. I would guess that the problem is that reverse iterators refer to the next element internally, and so the pre-incremented reverse iterator is still invalidated. I don't understand the problem with the implementation of the forward iterator version. As far as I can tell it's just one line shorter. Code for reference:
https://gist.github.com/jaj22/c89d213ec44c242f54c8016538b7b1fd

As far as I can tell, MSVC is correctly interpreting the C++ spec by throwing the errors. It's more that the C++ spec isn't well aligned with how vectors are implemented, so you'd never get a real crash from a wide variety of technically incorrect code.

I also get an unrelated crash in l_pigui_radial_menu, by opening the radial menu. It's a lua_istable assertion caused by the line `LuaTable xuv0(l, -1);`. I'm guessing that this was overlooked in the ImVec2 replacement.

Reply:
> Crashes.
😤

> I don't understand the problem with the implementation of the forward iterator version

I thought "reverse iterators" were designed (and they exist) to _solve_ this kind of situation :/ ... And for readability: I feel it was easier to get the idea when you saw a "reverse iterator" going "normally" instead of a forward iterator going backward (though in C this happens often), but if there are so many "nitpick" then it is definitely not worth the effort.

Well, enough is enough and you write that code for me: thank you. If you didn't find the remaining error (`LuaTable xuv0(l, -1);`) I had already pushed everything: I will push if not tomorrow this weekend when I hopefully will solve that issue (...there aren't iterator around that line ;) ).